core/CMakeLists.txt

```diff
@@ -0,0 +1,135 @@
+##################################################
+# Build a custom protoc plugin
+set(PROTOC_PLUGIN ${PROJECT_NAME}_protoc_plugin)
+
+gz_add_executable(${PROTOC_PLUGIN}
+  generator/Generator.cc
+  generator/generator_main.cc)
+target_link_libraries(${PROTOC_PLUGIN}
+  protobuf::libprotoc
+  protobuf::libprotobuf)
+target_include_directories(${PROTOC_PLUGIN} PRIVATE ${PROTOBUF_INCLUDE_DIR})
+target_compile_features(${PROTOC_PLUGIN} PRIVATE ${GZ_CXX_11_FEATURES})
+
+if (UNIX)
+  target_link_libraries(${PROTOC_PLUGIN} pthread)
+endif()
+
+install(TARGETS ${PROTOC_PLUGIN} DESTINATION ${GZ_BIN_INSTALL_DIR})
+
+##################################################
+# Generate messages
+file(GLOB proto_files ${PROJECT_SOURCE_DIR}/proto/gz/msgs/*.proto)
+
+include(${PROJECT_SOURCE_DIR}/cmake/gz_msgs_factory.cmake)
+include(${PROJECT_SOURCE_DIR}/cmake/gz_msgs_generate.cmake)
+include(${PROJECT_SOURCE_DIR}/cmake/gz_msgs_protoc.cmake)
+include(${PROJECT_SOURCE_DIR}/cmake/gz_msgs_string_utils.cmake)
+
+gz_msgs_generate_messages_impl(
```

Comment:
Can we not just use `gz_msgs_generate_messages_lib` here?

Reply:
Unfortunately not.  `gz_msgs_generate_messages_lib` needs to link to the `gz-msgs` lib.  That's why I introduced a new function that does most of the work _except_ actually creating the lib.

For gz-msgs, the message definitions are part of `gz-msgs.so`.  For other places, they will be in eg `gz-sim-msgs.so` which will then link to gz-msgs.so.

Reply:
I see, so `gz_msgs_generate_messages_lib` is for other libraries to use for custom messages, and in the future, per-package messages?

Reply:
Correct